uberfire-project/uberfire-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/POMServiceImpl.java

```diff
@@ -153,74 +163,118 @@ public Path save(final Path path,
-                content.getModules() != null) {
-            for (String module : content.getModules()) {
-
-                org.uberfire.java.nio.file.Path childPath = Paths.convert(path).getParent().resolve(module).resolve("pom.xml");
-
-                if (ioService.exists(childPath)) {
-                    POM child = load(Paths.convert(childPath));
-                    if (child != null) {
-                        child.setParent(content.getGav());
-                        child.getGav().setGroupId(content.getGav().getGroupId());
-                        child.getGav().setVersion(content.getGav().getVersion());
-
-                        save(Paths.convert(childPath),
-                             child);
-                    }
-                }
+        private void savePOM(final Path path,
+                             final POM content,
+                             final Metadata metadata) throws IOException, XmlPullParserException {
+            final Optional<Module> oldModule = getModuleIfNameChanged(path,
+                                                                      content);
+
+            if (metadata == null) {
+                ioService.write(Paths.convert(path),
+                                pomContentHandler.toString(content,
+                                                           loadPomXMLString(path)));
+            } else {
+                ioService.write(Paths.convert(path),
+                                pomContentHandler.toString(content,
+                                                           loadPomXMLString(path)),
+                                metadataService.setUpAttributes(path,
+                                                                metadata));
+            }
+
+            if (oldModule.isPresent()) {
+                updates.add(() -> renameModuleEvent.fire(new RenameModuleEvent(oldModule.get(),
+                                                                               moduleService.resolveModule(path))));
             }
         }
-    }
 
-    private Path save(final Path path,
-                      final POM content,
-                      final Metadata metadata) throws IOException, XmlPullParserException {
-        if (metadata == null) {
-            save(path,
-                 content);
-        } else {
-            ioService.write(Paths.convert(path),
-                            pomContentHandler.toString(content,
-                                                       loadPomXMLString(path)),
-                            metadataService.setUpAttributes(path,
-                                                            metadata));
+        private Optional<Module> getModuleIfNameChanged(final Path path,
+                                                        final POM content) {
+            if (!load(path).getName().equals(content.getName())) {
+                return Optional.of(moduleService.resolveModule(path));
```

Comment:
This seems incorrect to me (or I simply misunderstand how it works). You're calling `moduleService.resolveModule(path)` here and returning it in Optional. And then on line 241 you're creating RenameModuleEvent **with the same path** `moduleService.resolveModule(path)` (the path didn't change along the way). Is there some magic involved or are you actually passing the same 'old module' twice to RenameModuleEvent constructor?

Also renaming all occurrences of `path` to `pomPath` in this entire class (where it applies) would improve readability..

Reply:
The path is the same, but the content has been updated. So the two modules are different. Old is what it was before update and the one loaded on line 241 has the latest content.